huaweicloud/services/dws/data_source_huaweicloud_dws_logical_cluster_rings.go

```diff
@@ -0,0 +1,276 @@
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the DWS cluster ID.`,
+			},
+			"cluster_rings": {
+				Type:        schema.TypeList,
+				Elem:        clusterRingsSchema(),
+				Computed:    true,
+				Description: `Indicates the cluster ring list information.`,
+			},
+		},
+	}
+}
+
+func clusterRingsSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"is_available": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: `Indicates whether the cluster host ring is available.`,
+			},
+			"ring_hosts": {
+				Type:        schema.TypeList,
+				Elem:        ringHostsSchema(),
+				Computed:    true,
+				Description: `Indicates the cluster host ring list information.`,
+			},
+		},
+	}
+	return &sc
+}
+
+func ringHostsSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"host_name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `Indicates the host name.`,
+			},
+			"back_ip": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `Indicates the backend IP address.`,
+			},
+			"cpu_cores": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: `Indicates the number of CPU cores.`,
+			},
+			"memory": {
+				Type:        schema.TypeFloat,
+				Computed:    true,
+				Description: `Indicates the host memory.`,
+			},
+			"disk_size": {
+				Type:        schema.TypeFloat,
+				Computed:    true,
+				Description: `Indicates the host disk size.`,
+			},
+		},
+	}
+	return &sc
+}
+
+func readLogicalClusters(client *golangsdk.ServiceClient, d *schema.ResourceData) (interface{}, error) {
+	getPath := client.Endpoint + "v2/{project_id}/clusters/{cluster_id}/logical-clusters"
+	getPath = strings.ReplaceAll(getPath, "{project_id}", client.ProjectID)
+	getPath = strings.ReplaceAll(getPath, "{cluster_id}", d.Get("cluster_id").(string))
+	getResp, err := pagination.ListAllItems(
+		client,
+		"offset",
+		getPath,
+		&pagination.QueryOpts{MarkerField: ""})
+
+	if err != nil {
+		return nil, err
+	}
+
+	getRespJson, err := json.Marshal(getResp)
+	if err != nil {
+		return nil, err
+	}
+	var getRespBody interface{}
+	err = json.Unmarshal(getRespJson, &getRespBody)
+	if err != nil {
+		return nil, err
+	}
+	return getRespBody, nil
+}
+
+func readLogicalClusterRings(client *golangsdk.ServiceClient, d *schema.ResourceData) (interface{}, error) {
+	getPath := client.Endpoint + "v2/{project_id}/clusters/{cluster_id}/logical-clusters/rings"
+	getPath = strings.ReplaceAll(getPath, "{project_id}", client.ProjectID)
+	getPath = strings.ReplaceAll(getPath, "{cluster_id}", d.Get("cluster_id").(string))
+	getResp, err := pagination.ListAllItems(
+		client,
+		"offset",
+		getPath,
+		&pagination.QueryOpts{MarkerField: ""})
+
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving DWS logical cluster rings: %s", err)
```

Comment:
```suggestion
		return nil, err
```
it is better to be consistent with line 114, and the description is repeated with the invoke method

Reply:
OK